src/app/clusters/thread-border-router-management-server/thread-border-router-management-server.cpp

```diff
@@ -0,0 +1,384 @@
+        currentActiveDataset.GetActiveTimestamp(currentActiveDatasetTimestamp) == CHIP_NO_ERROR)
+    {
+        return Status::InvalidInState;
+    }
+
+    mBreadcrumb = req.breadcrumb;
+    mSetActiveDatasetSequenceNumber++;
+    CHIP_ERROR err = mDelegate->SetActiveDataset(activeDataset, mSetActiveDatasetSequenceNumber, this);
+    return StatusIB(err).mStatus;
+}
+
+Status ServerInstance::HandleSetPendingDatasetRequest(const Commands::SetPendingDatasetRequest::DecodableType & req)
+{
+    ReturnErrorCodeIf(!mDelegate, Status::InvalidInState);
+    bool panChangeSupported;
+    if (mDelegate->GetPanChangeSupported(panChangeSupported) != CHIP_NO_ERROR || !panChangeSupported)
+    {
+        return Status::UnsupportedCommand;
+    }
+    Thread::OperationalDataset pendingDataset;
+    // If any of the parameters in the PendingDataset is invalid, the command SHALL fail with a status code
+    // of INVALID_COMMAND.
+    ReturnErrorCodeIf(pendingDataset.Init(req.pendingDataset) != CHIP_NO_ERROR, Status::InvalidCommand);
+    CHIP_ERROR err = mDelegate->SetPendingDataset(pendingDataset);
+    return StatusIB(err).mStatus;
+}
+
+void AddDatasetResponse(CommandHandlerInterface::HandlerContext & ctx, Status status, Thread::OperationalDataset & dataset)
+{
+    if (status != Status::Success)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+    Commands::DatasetResponse::Type response;
+    response.dataset = dataset.AsByteSpan();
+    ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+}
+
+void ServerInstance::InvokeCommand(HandlerContext & ctxt)
+{
+    switch (ctxt.mRequestPath.mCommandId)
+    {
+    case Commands::GetActiveDatasetRequest::Id:
+        HandleCommand<Commands::GetActiveDatasetRequest::DecodableType>(ctxt, [this](HandlerContext & ctx, const auto & req) {
+            Thread::OperationalDataset dataset;
+            Status status = HandleGetActiveDatasetRequest(IsOverCASESession(ctx), dataset);
+            AddDatasetResponse(ctx, status, dataset);
+        });
+        break;
+    case Commands::GetPendingDatasetRequest::Id:
+        HandleCommand<Commands::GetPendingDatasetRequest::DecodableType>(ctxt, [this](HandlerContext & ctx, const auto & req) {
+            Thread::OperationalDataset dataset;
+            Status status = HandleGetPendingDatasetRequest(IsOverCASESession(ctx), dataset);
+            AddDatasetResponse(ctx, status, dataset);
+        });
+        break;
+    case Commands::SetActiveDatasetRequest::Id:
+        HandleCommand<Commands::SetActiveDatasetRequest::DecodableType>(ctxt, [this](HandlerContext & ctx, const auto & req) {
+            if (!mAsyncCommandHandle.Get())
+            {
+                mPath               = ctx.mRequestPath;
+                mAsyncCommandHandle = CommandHandler::Handle(&ctx.mCommandHandler);
+                ctx.mCommandHandler.FlushAcksRightAwayOnSlowCommand();
```

Comment:
This will not actually immediately flush acks, and could be omitted. That's because it's not until `HandleSetActiveDatasetRequest` is returns that the event loop will run. Perhaps the flush acks should be done in  `HandleSetActiveDatasetRequest` if it's known it will be doing async work.

Reply:
Done